### PR TITLE
[dune] Build coqc.byte executable.

### PR DESCRIPTION
**Pull request by ejgallego**
This may be useful in a few cases, like testing compilation with
byte-plugins; I chose to install it globally tho it is more of a
developer target.

### CODE REVIEW

File: topbin/dune
Comment (rgrinberg) [inline topbin/dune]: Isn't having the .bc target enough? If want to install it, then feel free to add an appropriate `(install ..)` stanza

Comment (ejgallego) [reply topbin/dune]: Indeed thanks, I dunno what I was thinking....

**Review by Zimmi48 — COMMENTED** [review]
It seems that this is reverting the change from #229 and #709. I don't have anything against it but this would be good to give a justification other than just "the maintainer of the build system is not the same person anymore".

**Comment by ejgallego** [comment]
> It seems that this is reverting the change from #229 and #709. I don't have anything against it but this would be good to give a justification other than just "the maintainer of the build system is not the same person anymore".

Sorry @Zimmi48 I am not sure how are you reading this PR, but it certainly has nothing to do (at all!) with the PRs you mention. The setup discussed there is just not relevant to a Dune build as it works in a different way w.r.t. byte / opt stuff. Of course feel free to ask any question.

This PR is very trivial in that it will just build one more executable and install it, doesn't have any other impact [you can compare the install files]

**Comment by Zimmi48** [comment]
I'm not blocking the PR and I can merge it without any delay, but the fact that the two build systems do not work in the same way doesn't change anything to the fact that from the user's perspective we'll have gone from a situation where both `coqc` and `coqc.byte` were installed, to a situation where only the former was installed, back to a situation where both are installed. The intermediate step was probably a mistake but this is nonetheless interesting to note how these decisions are taken...

**Comment by maximedenes** [comment]
> I am not sure how are you reading this PR, but it certainly has nothing to do (at all!) with the PRs you mention

Of course it is related.

@Zimmi48 The decision of not building bytecode executables as part of the default `make` target was taken to avoid race conditions when calling `ocamlopt` and `ocamlc` concurrently. AFAIK, Dune does not suffer from these issues, due to its specific knowledge of the OCaml requirements. So making Dune build bytecode executables by default is not invalidating the previous decision, it is just taking advantage of an improvement and seems logical.

**Review by Zimmi48 — APPROVED** [review]

**Comment by ejgallego** [comment]
> The intermediate step was probably a mistake but this is nonetheless interesting to note how these decisions are taken...

Théo, this PR was in response to a comment in #9974 .

But indeed trust me this PR is not related to the PRs you cite folks, in the sense the build rules are identical _after_ and _before_, the thing that changes is that `coqc.byte` is installed now.

You can build `coqc.byte` today, try `dune build topbin/coqc.bc`.

I am still not sure there is a use case for installing `coqc.byte` tho.

**Comment by Zimmi48** [comment]
>Théo, this PR was in response to a comment in #9974.

Sometimes, it would be helpful to give that kind of information from the start. Note that it changes much in this case.

>the build rules are identical after and before, the thing that changes is that coqc.byte is installed now

I'm aware of this. I do understand what the PR does. I'm talking about the user's perspective here.

>I am still not sure there is a use case for installing coqc.byte tho.

I don't know what I should conclude from this, given that this is all the PR does.

**Comment by maximedenes** [comment]
> Théo, this PR was in response to a comment in #9974 .

There you said:

> We don't even build it in Dune.

And now you seem to say the opposite...

**Comment by ejgallego** [comment]
> Sometimes, it would be helpful to give that kind of information from the start. Not that it changes much in this case.

Indeed, sorry.

> I'm talking about the user's perspective here.

Well indeed my point is that PRs you mentioned were not done on any user consideration, but on problems with the build itself, as Maxime said.

> I don't know what I should conclude from this, given that this is all the PR does.

Me neither, this PR is if you want a RFC, the outcome could well be that `coqc.byte` is not useful at all. But given how little it costs we can just ship it [well, it costs about 40 MiB (!!)]

**Comment by ejgallego** [comment]
> And now you seem to say the opposite...

We don't build it _by default_ I mean, the rules to build it are already there.

**Comment by maximedenes** [comment]
> We don't build it _by default_ I mean, the rules to build it are already there.

Ok, but then Théo was right that it was the same question as for the `Makefile`s. But I think that's clarified now.

**Comment by Zimmi48** [comment]
> > 
> 
> Well indeed my point is that PRs you mentioned were not done on any user consideration, but on problems with the build itself, as Maxime said.

I agree, and I think it was the main mistake at the time, to forget about the potential consequences for users. In the end, I think I have seen complaints from time to time about the absence of the byte outputs, so I'm still in favor of merging this PR if everybody agrees.

**Comment by ejgallego** [comment]
Just a nit, the PRs we mention talk about `coqtop` and `coqtop.byte`, note that the Dune build has installed both since the beginning; so in a sense, the decision that Théo talks about was taken a few months ago in an implicit way.

**Comment by Zimmi48** [comment]
Good point. And let's aim for consistency by doing the same thing for `coqc`. I will merge tomorrow if there are no barring comments.

**Pull request merged**
